File: Project/Simulator/Organisms/Organism.java
package Simulator.Organisms;

import Simulator.Tile;
import Simulator.World;
import Simulator.Organisms.Animals.*;
import Simulator.Organisms.Plants.*;

import org.json.JSONObject;

public abstract class Organism {
    protected int power;
    protected int initiative;
    protected long age = 0;
    protected boolean alive = true;
    protected int reproductionCooldown;
    protected boolean skip = false;
    protected Tile tile = null;
    protected World world;
    protected Type type;

    public enum Type {
        ANTELOPE("🦌"),
        // CYBER_SHEEP("🤖"),
        FOX("🦊"),
        HUMAN("👨"),
        SHEEP("🐑"),
        TURTLE("🐢"),
        WOLF("🐺"),
        GRASS("🌿"),
        GUARANA("🍅"),
        MILKWEED("🌾"),
        SOSNOWSKY_HOGWEED("🍁"),
        WOLF_BERRIES("🍇");

        private final String symbol;

        Type(String symbol) {
            this.symbol = symbol;
        }

        public String getSymbol() {
            return symbol;
        }

        public static Type getTypeByInt(int i) {
            return Type.values()[i];
        }

        public Organism construct(World w) {
            switch (this) {
                case ANTELOPE:
                    return new Antelope(w);
                case FOX:
                    return new Fox(w);
                case HUMAN:
                    return new Human(w);
                case SHEEP:
                    return new Sheep(w);
                case TURTLE:
                    return new Turtle(w);
                case WOLF:
                    return new Wolf(w);
                case GRASS:
                    return new Grass(w);
                case GUARANA:
                    return new Guarana(w);
                case MILKWEED:
                    return new Milkweed(w);
                case SOSNOWSKY_HOGWEED:
                    return new SosnowskyHogweed(w);
                case WOLF_BERRIES:
                    return new WolfBerries(w);
                default:
                    throw new IllegalStateException("Unexpected value: " + this);
            }
        }
    }

    public Type getType() {
        return type;
    }

    public Organism(Type type, int power, int initiative, World world) {
        this.type = type;
        this.power = power;
        this.initiative = initiative;
        this.world = world;
    }

    public Organism(JSONObject json, World world) {
        this.type = Type.getTypeByInt(json.getInt("type"));
        this.power = json.getInt("power");
        this.initiative = json.getInt("initiative");
        this.age = json.getLong("age");
        this.alive = json.getBoolean("alive");
        this.reproductionCooldown = json.getInt("reproduction_cooldown");
        this.skip = json.getBoolean("skip");
        this.world = world;
        this.tile = world.getTile(json.getInt("tile_index"));
    }

    public Tile getTile() {
        return tile;
    }

    public void setTile(Tile tile) {
        this.tile = tile;
    }

    public void age() {
        if (reproductionCooldown > 0)
            reproductionCooldown--;
        age++;
    }

    public void setBreedCooldown(int turns) {
        this.reproductionCooldown = turns;
    }

    public void die() {
        alive = false;
    }

    public boolean isDead() {
        return !alive;
    }

    public boolean isAlive() {
        return alive;
    }

    public void skipTurn() {
        skip = true;
    }

    public void unskipTurn() {
        skip = false;
    }

    public boolean isSkipped() {
        return skip;
    }

    public int getPower() {
        return power;
    }

    public int getInitiative() {
        return initiative;
    }

    public long getAge() {
        return age;
    }

    public void setPower(int power) {
        this.power = power;
    }

    public void setInitiative(int initiative) {
        this.initiative = initiative;
    }

    public String getSymbol() {
        return type.getSymbol();
    }

    public void draw() {
        System.out.println(getSymbol());
    }

    public abstract void action();

    public abstract void collision(Organism other);

    public abstract boolean collisionReaction(Organism other);

    public abstract Organism construct();

    public JSONObject toJson() {
        JSONObject json = new JSONObject();
        json.put("type", this.type.ordinal());
        json.put("power", this.power);
        json.put("initiative", this.initiative);
        json.put("age", this.age);
        json.put("alive", this.alive);
        json.put("reproduction_cooldown", this.reproductionCooldown);
        json.put("skip", this.skip);
        json.put("tile_index", this.tile.index);
        return json;
    }

    // public Organism(Map<String, Object> json, World world) {
    // // Assuming ObjectMapper can handle this conversion. You might need to
    // manually
    // // parse fields.
    // ObjectMapper mapper = new ObjectMapper();
    // Organism organism = mapper.convertValue(json, Organism.class);
    // this.type = organism.type;
    // this.power = organism.power;
    // this.initiative = organism.initiative;
    // this.age = organism.age;
    // this.alive = organism.alive;
    // this.reproductionCooldown = organism.reproductionCooldown;
    // this.skip = organism.skip;
    // this.tile = organism.tile;
    // this.world = world;
    // }
}

File: Project/Simulator/Organisms/Organism.py
from abc import ABC, abstractmethod

from enum import Enum


class Type(Enum):
    ANTELOPE = "🦌"
    CYBER_SHEEP = "🤖"
    FOX = "🦊"
    HUMAN = "👨"
    SHEEP = "🐑"
    TURTLE = "🐢"
    WOLF = "🐺"
    GRASS = "🌿"
    GUARANA = "🍅"
    MILKWEED = "🌾"
    SOSNOWSKY_HOGWEED = "🍁"
    BELLADONNA = "🍇"

    def getSymbol(self):
        return self.value

    def getName(self):
        return self.name.replace('_', ' ')

    def getTypeByInt(i):
        return list(Type)[i]


class Organism(ABC):
    def __init__(self, type: Type, power, initiative, world, json=None):
        if json is None:
            self._type = type
            self._power = power
            self._initiative = initiative
            self._age = 0
            self._alive = True
            self._reproductionCooldown = 0
            self._skip = False
            self._world = world
            self._tile = None
        else:
            self._type = Type.getTypeByInt(json["type"])
            self._power = json["power"]
            self._initiative = json["initiative"]
            self._age = json["age"]
            self._alive = json["alive"]
            self._reproductionCooldown = json["reproduction_cooldown"]
            self._skip = json["skip"]
            self._world = world
            self._tile = world.getTile(json["tile_index"])

    def getType(self):
        return self._type

    def getTile(self):
        return self._tile

    def setTile(self, tile):
        self._tile = tile

    def Age(self):
        if (self._reproductionCooldown > 0):
            self._reproductionCooldown -= 1
        self._age += 1

    def setBreedCooldown(self, turns):
        self._reproductionCooldown = turns

    def die(self):
        self._alive = False

    def isDead(self):
        return not self._alive

    def isAlive(self):
        return self._alive

    def skipTurn(self):
        self._skip = True

    def unskipTurn(self):
        self._skip = False

    def isSkipped(self):
        return self._skip

    def getPower(self):
        return self._power

    def getInitiative(self):
        return self._initiative

    def getAge(self):
        return self._age

    def setPower(self, power):
        self._power = power

    def setInitiative(self, initiative):
        self._initiative = initiative

    def getSymbol(self):
        return self._type.value

    def toJson(self):
        json = {
            "type": list(Type).index(self._type),
            "power": self._power,
            "initiative": self._initiative,
            "age": self._age,
            "alive": self._alive,
            "reproduction_cooldown": self._reproductionCooldown,
            "skip": self._skip,
            "tile_index": self._tile.INDEX if self._tile else None
        }
        return json

    @abstractmethod
    def action(self):
        pass

    @abstractmethod
    def collision(self, other):
        pass

    @abstractmethod
    def collisionReaction(self, other):
        pass

    @abstractmethod
    def construct(self):
        pass
